packages/statemanager/src/stateManager.ts

```diff
@@ -122,6 +122,11 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
     })
   }
 
+  async deleteAccount(address: Address) {
+    delete this._storageTries[address.buf.toString('hex')]
```

Comment:
In the case that a contract gets selfdestructed, this method gets called. However, before this change, it would not clear the storage trie cache. This means that while the contract should have empty storage, StateManager thinks the storage has not been wiped.

Does this lead to a consensus bug if you selfdestruct a CREATE2 contract and then CREATE2 the contract again? If EVM reads storage, does this contract then have storage or is it empty? It actually is, if we create a contract we ensure we wipe the storage first.

Reply:
The problem was this:

- Create a contract, which sets storage slot 1 to value 1. Selfdestruct this contract
- If you `getAccount` on the contract, it correctly returns an account which has nonce 0, balance 0, codeHash empty, storage empty
- However, if you `getContractStorage` on that address, and that key, it will return `Buffer < 01 >`. This is because the storage trie cache was not deleted, so storage trie still points to the contract storage state right before the selfdestruct.

This thus means that in some very exotic situations, if test runners (ganache, hardhat) try to test certain expected values, StateManager thus returns the wrongly cached values (so if you inspect StateManager it reports the wrong values). You could thus write a test which will not match the expected values due to not deleting the cached trie.